example/weather/services/tester/run_tests.go

```diff
@@ -0,0 +1,155 @@
+package tester
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime/debug"
+	"strings"
+	"sync"
+	"time"
+
+	gentester "goa.design/clue/example/weather/services/tester/gen/tester"
+	"goa.design/clue/log"
+	"golang.org/x/exp/slices"
+)
+
+// Ends a test by calculating duration and appending teh results to the test collection
+func endTest(tr *gentester.TestResult, start time.Time, tc *TestCollection, results []*gentester.TestResult) {
+	elapsed := time.Since(start).Milliseconds()
+	tr.Duration = elapsed
+	results = append(results, tr)
+	tc.AppendTestResult(results...)
+}
+
+// recovers from a panicked test. This is used to ensure that the test
+// suite does not crash if a test panics.
+func recoverFromTestPanic(ctx context.Context, testName string, testCollection *TestCollection) {
+	if r := recover(); r != nil {
+		msg := fmt.Sprintf("[Panic Test]: %v", testName)
+		err := errors.New(msg)
+		log.Errorf(ctx, err, fmt.Sprintf("%v", r))
+		// This doesn't work as I'd like because it only prints to stderr, not to a string
+		// 	that I can caputre and put in the goa log.
+		debug.PrintStack()
+		resultMsg := fmt.Sprintf("%v | %v", msg, r)
+		testCollection.AppendTestResult(&gentester.TestResult{
+			Name:     testName,
+			Passed:   false,
+			Error:    &resultMsg,
+			Duration: -1,
+		})
+	}
+}
+
+// Runs the tests from the testmap and handles filtering/exclusion of tests
+func (svc *Service) runTests(ctx context.Context, p *gentester.TesterPayload, testCollection *TestCollection, testMap map[string]func(context.Context, *TestCollection)) (res *gentester.TestResults, err error) {
+	retval := gentester.TestResults{}
+
+	testsToRun := make(map[string]func(context.Context, *TestCollection))
+	filtered := false
+	// If there are no filters, the tests to run is all tests in the map.
+	if (p == nil) || (p.Include == nil || len(p.Include) == 0) && (p.Exclude == nil || len(p.Exclude) == 0) {
+		testsToRun = testMap
+	} else { // Otherwise, we need to filter the tests
+		// If there is an include list, we only run the tests in the include list. This will supersede any exclude list.
+		filtered = true
+		if len(p.Include) > 0 {
+			if len(p.Exclude) > 0 {
+				return nil, gentester.MakeIncludeExcludeBoth(errors.New("cannot have both include and exclude lists"))
+			}
+			for _, test := range p.Include {
+				if testFunc, exists := testMap[test]; exists {
+					testsToRun[test] = testFunc
+				} else {
+					// QUESTION: Do we want to error the test execution if a test is not found in the test map?
+					// 		I'm thinking no, because it's not really an error, it's just a test that doesn't exist.
+					log.Infof(ctx, "Test [%v] not found in test map", test)
```

Comment:
@raphael  I was hesitant to even log an error because that could then show up in the Grafana dashboards as an error in the service when really it was a sort of user error for trying to filter on a test that doesn't exist. I think even with an Infof log, we could build a specific panel on a Tester dashboard to track when this happens if we really want to. 🤷🏻

Reply:
Yeah I don't see the error showing up in Grafana as a bad thing. At the end of the day someone needs to realize they've setup tests incorrectly and it's probably fine to expect users of the tester service to be looking at its error logs. An alternative could be to return a special test failure that indicates that the test wasn't found.